wagtail/blocks/stream_block.py

```diff
@@ -460,6 +460,49 @@ def insert(self, i, item):
         def __repr__(self):
             return repr(list(self))
 
+    class lazy_callable_stream(Mapping):
+        """
+        Internal helper used by the `blocks_by_name` method to allow
+        lazy evaluation of the blocks retrieved.
+        """
+
+        def __init__(self, block):
+            self.block = block
+            self.block_names = [
+                block_name for block_name in block.stream_block.child_blocks
+            ]
+
+        def __call__(self, block_name):
```

Comment:
I think we can change `blocks_by_name` to do an ordinary dict lookup if `block_name` is passed, and then we don't need this - that way we can make this object behave more like a conventional dict.